src/analysis/souffle/epsilon_analysis.dl

```diff
@@ -0,0 +1,77 @@
+#ifndef SRC_ANALYSIS_SOUFFLE_EPSILON_ANALYSIS_DL_
+#define SRC_ANALYSIS_SOUFFLE_EPSILON_ANALYSIS_DL_
+
+#include "src/analysis/souffle/sensitivity_analysis.dl" 
+
+#define GLOBAL_EPSILON_OPERATOR as("global_epsilon", Operator)
+#define PRIVACY_MECHANISM_OPERATOR as("privacy_mechanism", Operator)
+
+// Global epsilon is a value determined by users of Raksha. 
+// We could eventually set one as the default. 
+.decl isGlobalEpsilon(epsilon: number)
+isGlobalEpsilon(epsilon) :- isOperation(operation), 
+    operationHasOperator(operation, GLOBAL_EPSILON_OPERATOR),
+    operationHasAttribute(operation, ["epsilon", $NumberAttributePayload(epsilon)]).
+
+// We track every mechanism that adds noise. 
+// Right now, we are not tracking the type of noise (e.g. via Laplace of Gaussian mechanisms).
+.decl isDiffPrivacyMechanism(operation: Operation)
+isDiffPrivacyMechanism(operation) :- isOperation(operation), 
+    operationHasOperator(operation, PRIVACY_MECHANISM_OPERATOR). 
+
+// The epsilon of each privacy mechanism is tracked and later attached to values. 
+.decl diffPrivacyMechanismHasEpsilon(operation: Operation, epsilon: number)
+diffPrivacyMechanismHasEpsilon(operation, epsilon) :- isDiffPrivacyMechanism(operation),
+    operationHasAttribute(operation, ["epsilon", $NumberAttributePayload(epsilon)]).
+
+// We propagate privacy mechanisms backwards to inputs as per the PINQ paper.
+// This relation tracks every input to its (direct and indirect) outputs.   
+.decl isSqlPath(sql_input: AccessPath, sql_output: AccessPath)
+
+isSqlPath(sql_input, sql_output):- isSqlOperation(operation), 
+    isSqlInput(operation, sql_input),
+    operationHasResult(operation, sql_output); 
+    isDiffPrivacyMechanism(operation), 
+    operationHasOperandList(operation, input_list), 
+    flatten(input_list, sql_input),
+    operationHasResult(operation, sql_output).  
+
+isSqlPath(sql_input, sql_output):- isSqlOperation(operation), 
+    isSqlInput(operation, sql_input), 
+    operationHasResult(operation, intermediate_output), 
+    isSqlPath(intermediate_output, sql_output);
+    isDiffPrivacyMechanism(operation), 
+    operationHasOperandList(operation, input_list), 
+    flatten(input_list, sql_input), 
+    isSqlPath(sql_input, intermediate_output),
+    isSqlPath(intermediate_output, sql_output). 
```

Comment:
Is this missing the recursive part of the definition or am I failing to see it? Also, why would writing the relation like this simplify things? I feel like it would add unnecessary code.

Reply:
No, you're not failing to see it, I wrote it wrong. Trying again:

```
isSqlPath(sql_input, sql_input) :- isSqlInput(operation, sql_input).
isSqlPath(diff_priv_input, diff_priv_input) :- isDiffPrivacyMechanism(operation), 
    operationHasOperandList(operation, input_list), 
    flatten(input_list, sql_input).
    
isSqlPath(sql_input, sql_output) :-
  isSqlPath(sql_input, intermediate),
  operationHasOperandList(operation, input_list), flatten(input_list, intermediate),
  operationHasResult(operation, sql_output).
```

Here's what I'm trying to do: in the original code, I see you handling two cases, a case involving a sql operation and a case involving a differential privacy mechanism. It looks to me like the rules generated are entangling two separate concerns:

1. Starting a SQL path using a differential privacy mechanism and/or SQL operation.
2. Traversing a SQL path across an operation.

I'm trying to separate those concerns by just introducing a zero-length path for the initial case and then uniformly handling the paths afterwards. I may be misunderstanding your propagation code, though, so if this does not work, do feel free to keep what you have.

Reply:
Yes, that makes a lot of sense. I changed the code.